### PR TITLE
templates: Add f-string based templating implementation

### PR DESCRIPTION
**Pull request by chyroc**
Supported f-string and jinja2 template in python langchain. This PR aims to add f-string support to go langchain.
https://github.com/langchain-ai/langchain/blob/81886ad34598b2092ee9efd84e837c22c5faea58/libs/core/langchain_core/prompts/string.py#L86

Although f-string is a feature of python, within our company, our existing projects are based on python langchain, and a large number of f-string templates are stored in the database. 
We need to support f-string in go. In order to migrate existing business from python to go.

### CODE REVIEW

**Comment by chyroc** [comment]
> Can you clarify what this is for? Is there an issue explaining the use cases and the problem(s) this PR is solving?

support by langchain python: https://github.com/langchain-ai/langchain/blob/81886ad34598b2092ee9efd84e837c22c5faea58/libs/core/langchain_core/prompts/string.py#L86

File: prompts/internal/fstring/fstring.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package fstring
+
+func Format(format string, values map[string]any) (string, error) {
```
Comment (tmc) [inline prompts/internal/fstring/fstring.go:3]: please comment exported symbols

File: prompts/internal/fstring/fstring.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package fstring
```
Comment (tmc) [inline prompts/internal/fstring/fstring.go:1]: include a doc.go in every new package with a package-level documentation string

File: prompts/internal/fstring/parser.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+			continue
+		}
+
+		tmp, err = r.scanToRightCurlyBracket()
+		if err != nil {
+			return err
+		}
+		valName := strings.TrimSpace(string(tmp))
+		if valName == "" {
+			return fmt.Errorf("empty expression not allowed")
```
Comment (tmc) [inline prompts/internal/fstring/parser.go:45]: please heed this lint warning and define an exported error value fo these conditions

**Review by tmc — COMMENTED** [review]
Please update the PR description with a description of the change and fix the lint issues.

I'm a little bit skeptical taking on our own f-string implementation (vs something that is already out there in Go) but am open to it.

**Comment by eliben** [comment]
> > Can you clarify what this is for? Is there an issue explaining the use cases and the problem(s) this PR is solving?
> 
> support by langchain python: https://github.com/langchain-ai/langchain/blob/81886ad34598b2092ee9efd84e837c22c5faea58/libs/core/langchain_core/prompts/string.py#L86

f-string is a Python feature and is useful and natural in Python code. It's not something that exists in Go -- perhaps a `text/template`-based implementation would be more natural.

**Comment by tmc** [comment]
> > > Can you clarify what this is for? Is there an issue explaining the use cases and the problem(s) this PR is solving?
> > 
> > 
> > support by langchain python: https://github.com/langchain-ai/langchain/blob/81886ad34598b2092ee9efd84e837c22c5faea58/libs/core/langchain_core/prompts/string.py#L86
> 
> f-string is a Python feature and is useful and natural in Python code. It's not something that exists in Go -- perhaps a `text/template`-based implementation would be more natural.

We already have a text/template based templating scheme and just recently added a jinja2-inspired approach. While I don't think we should have a large number of templating implementations I'm open to alternatives to some extent. Happy to have the discussion.

**Comment by chyroc** [comment]
@tmc update the PR description, updated the purpose and background of this PR

**Review by tmc — APPROVED** [review]
LGTM

**Pull request merged**